src/mscorlib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -777,6 +776,76 @@ public int EnsureCapacity(int capacity)
             return newSize;
         }
 
+        /// <summary>
+        /// Sets the capacity of this dictionary to what it would be if it had been originally initialized with all its entries
+        /// 
+        /// This method can be used to minimize the memory overhead 
+        /// once it is known that no new elements will be added. 
+        /// 
+        /// To completely clear a dictionary and release all memory 
+        /// referenced by the dictionary, execute the following statements:
+        /// 
+        /// dictionary.Clear();
+        /// dictionary.TrimExcess();
+        /// </summary>
+        public void TrimExcess()
+        {
+            TrimExcess(HashHelpers.GetPrime(Count));
+        }
+
+        /// <summary>
+        /// Sets the capacity of this dictionary to hold up 'capacity' entries without any further expansion of its backing storage
+        /// 
+        /// This method can be used to minimize the memory overhead 
+        /// once it is known that no new elements will be added. 
+        /// </summary>
+        public void TrimExcess(int capacity)
+        {
+            if (capacity < Count)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.capacity);
+            int newSize = HashHelpers.GetPrime(capacity);
+            int currentCapacity = _entries == null ? 0 : _entries.Length;
+            if (newSize >= currentCapacity)
+                return;
+            if (newSize < _count)
+            {
+                ref Entry[] entriesRef = ref _entries;
+                // Update free list to only include unused entries with index smaller than newSize
+                int index = _freeList;
+                int freeList = -1;
+                int next;
+                while (index != -1)
+                {
+                    next = entriesRef[index].next;
+                    if (index < newSize)
+                    {
+                        entriesRef[index].next = freeList;
+                        freeList = index;
+                    }
+                    index = next;
+                }
+
+                int count = _count;
```

Comment:
Just by looking at number of lines in this implementation, it is hard to believe that this would be the leanest way to implement this.

I have tried to write the most straightforward implementation possible: 
https://github.com/jkotas/coreclr/commit/ed51476d5d26a9be56220cfc1bb3b6b1500f2e16 It is smaller code than this.

Reply:
> @jkotas the original approach adds 19Kb to IL size. the implementation can be seen in commit 00c6b91
> A slight modification to original approach (dd514cd) would reduce 19Kb down to 18Kb, which is still not better than Approach2.

Your implementation here is the best one so far and as you said it adds less than others to native code (14.5Kb).
It's different from Approach2 (16Kb). I remember trying this (dd514cd) but since I was also doing an unnecessary Array.Copy prior to initialize, the code size ended up being higher (18Kb) and then I abandoned improving it further.

I'll update the PR to your approach right away.

Reply:
The purpose of version is to flag when enumeration is invalidated. That would happen iff the non-zombie entries in _entries are changed or re-ordered. So it seems to me that any operation which may do this, should version++.

I suppose there is value in flagging unsynchronized access to the object even if it does not change enumeration order but I'm guessing that isn't why version exists.

Reply:
I see now. In this case then I think it would have still been required in Approach2. Thanks for your explanation @danmosemsft